### ISSUE TITLE
[studio-ui] Password requirements should be displayed when password requirements is enabled

### ISSUE THREAD

**Issue by alhambrav** [craftercms/craftercms#3530]
## Describe the bug
Password requirements should be displayed when password requirements is enabled when creating a new user or changing password.  It would be nice too to have the password checked for requirements in the `Password` field instead of after clicking on the `Save`/`Submit` button

## To Reproduce
1. Startup Studio
2. Click on `Global Config` and add in the following to enable password requirements:
```
# Enable password requirements validation
studio.security.passwordRequirements.enabled: true
# Password requirements validation regular expression
#   (?=.*[0-9])       a digit must occur at least once
#   (?=.*[a-z])       a lower case letter must occur at least once
#   (?=.*[A-Z])       an upper case letter must occur at least once
#   (?=.*[@#$%^&+=])  a special character must occur at least once
#   (?=\S+$)          no whitespace allowed in the entire string
#   .{8,}             anything, at least eight places though
studio.security.passwordRequirements.validationRegex: ^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&+=])(?=\S+$).{8,}$
# Password requirements description
studio.security.passwordRequirements.description: >
  Password requirements
    Must have at least one digit
    Must have at least one lower case letter
    Must have at least one upper case letter
    Must have at least one special character
    Can not have whitespaces
    Must be at least eight characters long
```
Click on `Save`
3. Click on `Users` from the `Main Menu` then click on `New User`
4. Fill in the `First Name`, `Last Name`, `Email`, and `Username` 
5. Start filling in the `Password` field.  It would be nice if the password requirements are displayed somewhere in the dialog if password requirements is enabled. Enter a password that does not fulfill the password requirements
6. Notice that the password is validated against the password requirements after clicking on the `Save` button.  In other apps when you create a login, the password is validated against the requirement in the first `Password` field, not in the field where you retype the password.  It would be nice if we did the same for creating a user.

![Screen Shot 2019-10-14 at 11 58 49 AM](https://user-images.githubusercontent.com/25483966/66765844-c03aa380-ee7a-11e9-926b-878ea11a4d73.png)

7. After clicking on `Save`, you will get the message that:
```
User password does not fullfill requirements. Use password that will fullfill password requirements
```
Please fix the typo for `fulfill`.  Also, it would be nice if the cursor goes to the `Password` field after displaying the message that password requirements were not met.

![Screen Shot 2019-10-14 at 12 47 12 PM](https://user-images.githubusercontent.com/25483966/66768642-0d217880-ee81-11e9-9918-1b4dd8c2ea67.png)

8. Logout as user `admin` and login as another user that has been created successfully.
9. Click on the username then `Settings` from the top right of Studio
10. Go to the `Change Password` section and fill in the `Current Password`.  Notice that it would be nice if the password requirements are listed here if it is enabled
11. Fill in the `New Password` and `Confirm Password` with something that does not fulfill the password requirements.
12. Click on `Submit`.  Please fix the typo for `fulfill`.  Also, it would be nice if the cursor goes to the `New Password` field after displaying the message that password requirements were not met.

![Screen Shot 2019-10-14 at 12 53 53 PM](https://user-images.githubusercontent.com/25483966/66769162-6dfd8080-ee82-11e9-9bba-5ce3d7f77dd8.png)

## Expected behavior
Password requirements should be displayed when password requirements is enabled. 
It would be nice too to have the password checked for requirements in the `Password` field instead of after clicking on the `Save` button

## Screenshots
{{If applicable, add screenshots to help explain your problem.}}

## Logs
{{If applicable, attach the logs/stack trace (use https://gist.github.com).}}

### Version
Studio Version Number: 3.1.4-SNAPSHOT-fef9e6
Build Number: fef9e63180f08f17b786d3db1cf369e747d46034
Build Date/Time: 10-14-2019 08:03:01 -0400

### OS
OS X

### Browser
Chrome browser

## Additional context
{{Add any other context about the problem here.}}

**Comment by sumerjabri** [comment]
@dejan-brkic please add an API spec for getting the configuration items required from Studio.
The API should be something like:
- `/configuration/get_configuration_item`
  - module
  - item
There should be a white-list of what's permissible to be read, please add that per our conversation.

Once the API spec is ready, send a PR for review and begin implementation.

**Comment by dejan-brkic** [comment]
Done
PR:
https://github.com/craftercms/studio/pull/1786
https://github.com/craftercms/studio-ui/pull/1345

**Comment by sumerjabri** [comment]
@vestrada the Regex is now available to you in the context. The help text will need to be moved to the localization bundles and the regex needs to be modified to include the lookup key. Roy can better explain this.

**Comment by rart** [comment]
@vestrada, we will be using capture groups to name and reference the pieces of the validation regex and dynamically generating a localized dynamic tooltip. Not alll browsers support capture groups see full _algorithm_ below.

1. Get the expression from config
2. Extract the capture groups
3. Test if JS is able to compile the expression
    3.1. If expression doesn't compile, using the _capture group name extraction expression_, remove capture groups and attempt to compile again.
   3.2. If at all not able to compile or capture groups not supported (e.g. firefox) the tooltip with password requirements would be static and actual validation differs to backend. UI only validates that password is supplied and it matches it's confirmation. 
4. If expression is able to be compiled in JS and capture groups are supported - which you may probe for by using a sample string and sample RegExp - then the tooltip becomes dynamic. Technically, if the expressing compiles, capture groups are available and you may not even need to probe. Do a little multi-browser test to determine.
5. The dynamic tooltip would keep a list of each piece of the regex and initially set all conditions to not passed, showing a red `X` or similar icon indicating rules are not passed. To be able to individually validate each piece of the regex separately, we need to split into individual regex objects and test the user input against each. To split the regex, use the _capture group extraction expression_.

----

New default RegExp:
```
^(?=(?<hasNumbers>.*[0-9]))(?=(?<hasLowercase>.*[a-z]))(?=(?<hasUppercase>.*[A-Z]))(?=(?<hasSpecialChars>.*[@#$%^&+=]))(?=(?<noSpaces>\S+$))(?<minMaxLength>.{8,})$
```

Capture group name extraction expression
```js
/\?<(.*?)>/g
```

Capture group extraction expression
```js
\\(\?<(.*?)>(.*?)\)/g
```

JS min/max extraction RegExps
```js
// Extracting minMax section
\(\?<minMaxLength>(.*){(.*?)}\)


// Extracting minMax - .split(',') the output, min is first, max is second.
/{(.*?)}/
```

**Comment by vestrada** [comment]
@dejan-brkic 

I need passwordRequirementsRegex property on entry.ftl. I tried to do it including envconfig to it. But in login page it is giving errors. So We will need to added it on a different way. 

Thanks.

**Comment by russdanner** [comment]
Related ticket:
https://github.com/craftercms/craftercms/issues/3314

Current API
- Does not return password requirements
- Throws error in logs on password that does not meet requirements. This is not an error and should not be logged as one.
- There is a spelling mistake in the current message "Fullfill" is spelled "Fulfill"

**Comment by dejan-brkic** [comment]
Done
PR:
https://github.com/craftercms/studio/pull/1808


@vestrada 
It is available on `entry.ftl` as model variable `passwordRequirementsRegex` e.g. `${passwordRequirementsRegex}`

**Comment by vestrada** [comment]
@dejan-brkic please fix type on error message. From fullfill to fulfill.

**Comment by dejan-brkic** [comment]
fixed with another ticket

**Comment by alhambrav** [comment]
Verified typo for `fulfill` is fixed.

**Comment by alhambrav** [comment]
@vestrada  @dejan-brkic (not sure who to ask for the global config)

The commented out configuration for password requirements in the global config and the sample code contains outdated `studio.security.passwordRequirements.validationRegex`.  
This is the `validationRegex` that I used (given to me by Vero):
```
studio.security.passwordRequirements.validationRegex: ^(?=(?<hasNumbers>.[0-9]))(?=(?<hasLowercase>.[a-z]))(?=(?<hasUppercase>.[A-Z]))(?=(?<hasSpecialChars>.[@#$%^&+=]))(?=(?<noSpaces>\S+$))(?<minLength>.{8,})$
```
Please update accordingly.  

Also, the Password requirements description `studio.security.passwordRequirements.description` does not seem to be used at all when displaying the password requirements.  Maybe remove it from the configuration?  


Here's the existing configuration:
```
# Enable password requirements validation
# studio.security.passwordRequirements.enabled: false
# Password requirements validation regular expression
#   (?=.*[0-9])       a digit must occur at least once
#   (?=.*[a-z])       a lower case letter must occur at least once
#   (?=.*[A-Z])       an upper case letter must occur at least once
#   (?=.*[@#$%^&+=])  a special character must occur at least once
#   (?=\S+$)          no whitespace allowed in the entire string
#   .{8,}             anything, at least eight places though
# studio.security.passwordRequirements.validationRegex: ^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&+=])(?=\S+$).{8,}$
# Password requirements description
# studio.security.passwordRequirements.description: >
#   Password requirements
#     Must have at least one digit
#     Must have at least one lower case letter
#     Must have at least one upper case letter
#     Must have at least one special character
#     Can not have whitespaces
```
@vestrada  @rart 

I have some comments about the displayed password requirements too.
![Screen Shot 2019-11-01 at 2 14 41 PM](https://user-images.githubusercontent.com/25483966/68046332-03d24180-fcb2-11e9-9e8c-bd94e7f00989.png)

What do you think about changing the wording for the requirements to something like this:

```
Must contain at least one number
Must contain at least one lowercase character
Must contain at least one uppercase character
Must contain at least one special character (if possible list the allowed special characters here)
Blank spaces aren't allowed
Must contain at least 8 characters (this is the minimum length)
```
Or something like that.  

Thank you.

**Comment by rart** [comment]
@alhambrav messages have been updated (pending merge of https://github.com/craftercms/studio-ui/pull/1399). Good catch!

**Comment by dejan-brkic** [comment]
Done
PR:
https://github.com/craftercms/craftercms/pull/3647
https://github.com/craftercms/studio/pull/1823

**Comment by alhambrav** [comment]
Hi! Just a few more issues encountered:

* The UI does not seem to be honoring `studio.security.passwordRequirements.enabled: false`.  After a fresh install, the default value is `false` for `studio.security.passwordRequirements.enabled`, but, when you create a user and type in a password, the password requirements are being shown

* The password requirements are not displayed when an admin resets a password
  
![Untitled](https://user-images.githubusercontent.com/25483966/68605026-b310d480-0479-11ea-91c6-c9c5d78bbfe0.png)
 
* The password requirements are not displayed when resetting password from a link in email (Forgot password link from login dialog)
![Screen Shot 2019-11-11 at 11 45 12 AM](https://user-images.githubusercontent.com/25483966/68605076-cde34900-0479-11ea-92a2-1725ef387698.png)

Thank you!

**Comment by vestrada** [comment]
Hi @dejan-brkic 
Please add: `studio.security.passwordRequirements.enabled `to corresponfing ftl (entry.ftl)

**Comment by sumerjabri** [comment]
@dejan-brkic and @vestrada I don't want the passwordRequirments.enabled flag anymore. I want it to be always enabled, and the admin can make it easier/harder via the regex.

**Comment by sumerjabri** [comment]
Regarding the passwordRequirements.enabled flag:

@dejan-brkic remove it from the config files and backend
@vestrada remove it from the front-end

**Comment by dejan-brkic** [comment]
Done
PR:
https://github.com/craftercms/studio/pull/1828

**Comment by alhambrav** [comment]
Verified fixed.

**Comment by rart** [comment]
Missing fallback to raw key when not in built-in translations.

**Comment by alhambrav** [comment]
Verified.
